elegy/metrics/reduce_confusion_matrix.py

```diff
@@ -0,0 +1,141 @@
+            # Broadcast weights if possible.
+            sample_weight = jnp.broadcast_to(sample_weight, y_true.shape)
+        except ValueError as e:
+            raise e
+
+    if reduction == Reduction.TRUE_POSITIVES:
+        if sample_weight is not None:
+            y_true = y_true * sample_weight
+            y_pred = y_pred * sample_weight
+        mask = y_pred == 1
+        value = jnp.sum(y_true[mask] == 1)
+
+    if reduction == Reduction.FALSE_POSITIVES:
+        if sample_weight is not None:
+            y_true = y_true * sample_weight
+            y_pred = y_pred * sample_weight
+        mask = y_pred == 1
+        value = jnp.sum(y_true[mask] == 0)
+
+    if reduction == Reduction.FALSE_NEGATIVES:
+        if sample_weight is not None:
+            y_true = y_true * sample_weight
+            y_pred = y_pred * sample_weight
+        mask = y_true == 1
+        value = jnp.sum(y_pred[mask] == 0)
+
+    if reduction == Reduction.TRUE_NEGATIVES:
+        if sample_weight is not None:
+            y_true = y_true * sample_weight
+            y_pred = y_pred * sample_weight
+        mask = y_true == 0
+        value = jnp.sum(y_pred[mask] == 0)
+
+    cm_metric += value
+    return cm_metric
+
+
+class ReduceConfusionMatrix(Metric):
+    """Encapsulates confusion matrix metrics that perform a reduce operation on the values."""
+
+    def __init__(
+        self, reduction: Reduction, on: tp.Optional[types.IndexLike] = None, **kwargs
+    ):
+        super().__init__(on=on, **kwargs)
+
+        self._reduction = reduction
+
+        if self._reduction not in (
+            Reduction.FALSE_NEGATIVES,
+            Reduction.FALSE_POSITIVES,
+            Reduction.TRUE_POSITIVES,
+        ):
+            raise NotImplementedError(
+                "reduction {reduction} not implemented".format(
+                    reduction=self._reduction
+                )
+            )
+
+    def call(
+        self,
+        y_true: jnp.ndarray,
+        y_pred: jnp.ndarray,
+        sample_weight: tp.Optional[jnp.ndarray] = None,
+    ) -> jnp.ndarray:
+        """
+        Accumulates confusion matrix metrics for computing the reduction metric.
+
+        Arguments:
+            y_true: Ground truth values. shape = `[batch_size, d0, .. dN]`.
+
+            y_pred: The predicted values. shape = `[batch_size, d0, .. dN]`.
+
+            sample_weight: Optional weighting of each example. Defaults to 1.
+
+        Returns:
+            Array with the cummulative reduce metric.
+        """
+
+        if self._reduction in (
+            Reduction.FALSE_NEGATIVES,
+            Reduction.FALSE_POSITIVES,
+            Reduction.TRUE_POSITIVES,
+            Reduction.TRUE_NEGATIVES,
+        ):
+            cm_metric = hooks.get_state(
+                self._reduction.value,
+                shape=[],
+                dtype=jnp.int32,
+                initializer=initializers.Constant(0),
+            )
+        else:
+            count = None
```

Comment:
`cm_metric` is undefined in this branch

Reply:
Actually `count` is never used, did you mean `cm_metric = None`?

Reply:
yes I mean cm_metric, fixing this bug